doc/adr/0101-separating-concerns-between-persisters-actions.md

```diff
@@ -0,0 +1,133 @@
+# 0101 Separating concerns between workflow persisters and actions
+
+Date: 2025-01-02
+
+## Status
+
+Accepted
+
+## Context
+
+The `Workflow` library provides a state machine engine which is used in
+LedgerSMB to manage [resource
+state](./0017-state-machines-for-resource-state-management.md). Other systems
+use state machine engines for execution of processes; this is not the primary
+purpose of `Workflow` in LedgerSMB, although it may - to some extent - be a
+valid consequence.
+
+The components in `Workflow` have specific roles. Since LedgerSMB's storage
+medium (the PostgreSQL database) isn't just passive storage, but includes
+stored procedures to become "active" as well, the allocation of
+responsibilities becomes blurred: should the persister be the sole component
+interacting with the database or should actions do so as well? Where to draw
+the lines?
+
+### Summary of the `Workflow` library
+
+States are grouped into workflow types. These workflow types serve as templates
+or classes which can be instantiated into actual workflows. These workflow
+instances are what the workflow engine operates on; they are serialized for
+storage and deserialized to be loaded back into the running application. In
+LedgerSMB, the storage layer is the PostgreSQL database.
+
+#### Components making up `Workflow`
+
+The library consists of the following components:
+
+* States  
+  Define associated actions, their associated conditions and their transition
+  target states
+* Conditions  
+  Define conditions (Perl code) for availability of actions
+* Actions  
+  Define which Perl code is run when the action is triggered -- this may include
+  running stored procedures in the PostgreSQL database
+* Validators  
+  Define checks to establish input validity in the workflow state before
+  executing the code associated with an action
```

Comment:
What's the difference between a validator and a condition? (is a distinction necessary?)

Reply:
Added documentation of the difference to the file. As for whether the distinction is necessary: they serve a different purpose. Even when we don't find the distinction important, they *are* components that the Workflow library provides (which is why I added them to the explanation).